cpp/src/groupby/sort/group_tdigest.cu

```diff
@@ -0,0 +1,849 @@
+ * to be added.
+ *
+ * The number of clusters generated will always be <= delta_, where delta_ is
+ * a reasonably small number likely << 10000.
+ *
+ * Each input group gets an independent set of clusters generated.
+ *
+ * @param delta_             tdigest compression level
+ * @param num_groups         The number of input groups
+ * @param nearest_weight     A functor which returns the nearest weight in the input
+ * stream that falls before our current cluster limit
+ * @param total_weight       A functor which returns the expected total weight for
+ * the entire stream of input values for the specified group.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ *
+ * @returns A tuple containing the set of cluster weight limits for each group, a set of
+ * list-style offsets indicating group sizes, and the total number of clusters
+ */
+template <typename TotalWeightIter, typename NearestWeight, typename CumulativeWeight>
+std::tuple<rmm::device_uvector<double>, std::unique_ptr<column>, size_type>
+generate_group_cluster_info(int delta,
+                            size_type num_groups,
+                            NearestWeight nearest_weight,
+                            TotalWeightIter total_weight,
+                            CumulativeWeight cumulative_weight,
+                            rmm::cuda_stream_view stream,
+                            rmm::mr::device_memory_resource* mr)
+{
+  constexpr size_type block_size = 256;
+  cudf::detail::grid_1d const grid(num_groups, block_size);
+
+  // compute number of clusters per group
+  // each thread computes 1 set of clusters (# of cluster sets == # of groups)
+  rmm::device_uvector<size_type> group_num_clusters(
+    num_groups, stream, rmm::mr::get_current_device_resource());
+  generate_cluster_limits_kernel<<<grid.num_blocks, block_size, 0, stream.value()>>>(
+    delta,
+    num_groups,
+    nearest_weight,
+    total_weight,
+    cumulative_weight,
+    nullptr,
+    group_num_clusters.begin(),
+    nullptr);
+
+  // generate group cluster offsets (where the clusters for a given group start and end)
+  auto group_cluster_offsets = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, num_groups + 1, mask_state::UNALLOCATED, stream, mr);
+  auto cluster_size = cudf::detail::make_counting_transform_iterator(
+    0, [group_num_clusters = group_num_clusters.begin(), num_groups] __device__(size_type index) {
+      return index == num_groups ? 0 : group_num_clusters[index];
+    });
+  thrust::exclusive_scan(rmm::exec_policy(stream),
+                         cluster_size,
+                         cluster_size + num_groups + 1,
+                         group_cluster_offsets->mutable_view().begin<offset_type>(),
+                         0);
+
+  // total # of clusters
+  offset_type total_clusters;
+  cudaMemcpyAsync(&total_clusters,
```

Comment:
`get_value`?

Reply:
Yah.  I think at one point, `group_cluster_offsets` was a `device_uvector`, which is why I had it that way.